### PR TITLE
Payload Events

### PR DESCRIPTION
**Pull request by loocapro**
Related to: https://github.com/paradigmxyz/reth/issues/6507

Added `PayloadEvents` and sending events on payload attributes and payload resolve.

@mattsse what should be the right way to send the ResolvePayloadFuture from the broadcast::Sender?

### CODE REVIEW

**Review by mattsse — CHANGES_REQUESTED** [review]
I see that we're returning a `PayloadFuture`:

https://github.com/paradigmxyz/reth/blob/17eca361e8951362f5c544a2dfe20c9b046bef5f/crates/payload/builder/src/service.rs#L24-L24

what we could do is write our own future type here that also gets a clone of the sender and sends it once resolved or keep using box by extending the async scope:

https://github.com/paradigmxyz/reth/blob/17eca361e8951362f5c544a2dfe20c9b046bef5f/crates/payload/builder/src/service.rs#L287-L291

with an update

```rust
let broadcast = self.broadcast.clone();
```



I also want a helper struct for the Subscription so it's easier to get a stream of PayloadAttributes or Payloads

File: crates/payload/builder/src/service.rs [hunk 1]
```diff
@@ -171,6 +170,14 @@ where
     ) -> Result<PayloadId, PayloadBuilderError> {
         self.send_new_payload(attr).await?
     }
+
+    /// Sends a message to the service to subscribe to payload events.
+    /// Returns a receiver that will receive the payload events.
+    pub fn subscribe(&self) -> oneshot::Receiver<broadcast::Receiver<PayloadEvents<Engine>>> {
```
Comment (mattsse) [inline crates/payload/builder/src/service.rs:176]: I want a wrapper struct for `broadcast::Receiver` here so we can easily add more functions, for example to get a stream of PayloadAttr

File: crates/payload/builder/src/traits.rs [hunk 2]
```diff
@@ -25,7 +25,7 @@ pub trait PayloadJob: Future<Output = Result<(), PayloadBuilderError>> + Send +
         + Sync
         + 'static;
     /// Represents the built payload type that is returned to the CL.
-    type BuiltPayload: BuiltPayload + std::fmt::Debug;
+    type BuiltPayload: BuiltPayload + Clone + std::fmt::Debug;
```
Comment (mattsse) [inline crates/payload/builder/src/traits.rs:28]: I think this is reasonable bound, shouldn't impact anything.

and I don't think the clone is that expensive

File: crates/payload/builder/src/service.rs [hunk 3]
```diff
@@ -472,15 +480,15 @@ pub enum PayloadServiceCommand<Engine: EngineTypes> {
 #[derive(Clone)]
 pub enum PayloadEvents<Engine: EngineTypes> {
     Attributes(Engine::PayloadBuilderAttributes),
-    Resolve(Arc<PayloadFuture<Engine::BuiltPayload>>),
+    BuiltPayload(Engine::BuiltPayload),
 }
```
Comment (mattsse) [inline crates/payload/builder/src/service.rs:484]: this needs some docs

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/payload/builder/src/service.rs [hunk 4]
```diff
@@ -477,12 +480,24 @@ pub enum PayloadServiceCommand<Engine: EngineTypes> {
...
     Attributes(Engine::PayloadBuilderAttributes),
+    /// The built payload that has been just built.
+    /// Triggered by the CL whenever it asks for an execution payload.
+    /// This event is only thrown if the CL is a validator.
     BuiltPayload(Engine::BuiltPayload),
 }
 
+/// Represents a receiver for various payload events.
+#[derive(Debug)]
+pub struct PayloadEventReceiver<Engine: EngineTypes> {
```
Comment (mattsse) [inline crates/payload/builder/src/service.rs:499]:
```suggestion
pub struct PayloadEvents<Engine: EngineTypes> {
```

Comment (mattsse) [inline crates/payload/builder/src/service.rs]: I want to move this to a new file events.rs

Comment (mattsse) [inline crates/payload/builder/src/service.rs]: this should expose an async function to recv

**Review by mattsse — CHANGES_REQUESTED** [review]
last nits

**Review by onbjerg — APPROVED** [review]

**Review by mattsse — APPROVED** [review]
lgtm

**Pull request merged**